ecosystem/indexer-grpc/indexer-grpc-utils/src/file_store_operator.rs

```diff
@@ -0,0 +1,324 @@
+        version: u64,
+    ) -> anyhow::Result<()> {
+        if (std::time::Instant::now() - self.latest_metadata_update_timestamp).as_secs() < 5 {
+            return Ok(());
+        }
+
+        let metadata = FileStoreMetadata::new(chain_id, version);
+        // If the metadata is not updated, the indexer will be restarted.
+        match Object::create(
+            self.bucket_name.as_str(),
+            serde_json::to_vec(&metadata).unwrap(),
+            METADATA_FILE_NAME,
+            JSON_FILE_TYPE,
+        )
+        .await
+        {
+            Ok(_) => {
+                self.latest_metadata_update_timestamp = std::time::Instant::now();
+                Ok(())
+            },
+            Err(err) => Err(anyhow::Error::from(err)),
+        }
+    }
+
+    /// Uploads the transactions to the file store. The transactions are grouped into batches of BLOB_STORAGE_SIZE.
+    /// Updates the file store metadata after the upload.
+    pub async fn upload_transactions(
+        &mut self,
+        chain_id: u64,
+        transactions: Vec<EncodedTransactionWithVersion>,
+    ) -> anyhow::Result<()> {
+        let start_version = transactions.first().unwrap().1;
+        let batch_size = transactions.len();
+        anyhow::ensure!(
+            start_version % BLOB_STORAGE_SIZE as u64 == 0,
+            "Starting version has to be a multiple of BLOB_STORAGE_SIZE."
+        );
+        anyhow::ensure!(
+            batch_size % BLOB_STORAGE_SIZE == 0,
+            "The number of transactions to upload has to be multiplier of BLOB_STORAGE_SIZE."
+        );
+        let mut tasks = vec![];
+
+        // Split the transactions into batches of BLOB_STORAGE_SIZE.
+        for i in transactions.chunks(BLOB_STORAGE_SIZE) {
+            let bucket_name = self.bucket_name.clone();
+            let current_batch = i.iter().cloned().collect_vec();
+            let transactions_file = build_transactions_file(current_batch).unwrap();
+            let task = tokio::spawn(async move {
+                match Object::create(
+                    bucket_name.clone().as_str(),
+                    serde_json::to_vec(&transactions_file).unwrap(),
+                    generate_blob_name(transactions_file.starting_version).as_str(),
+                    JSON_FILE_TYPE,
+                )
+                .await
+                {
+                    Ok(_) => Ok(()),
+                    Err(err) => Err(anyhow::Error::from(err)),
+                }
+            });
+            tasks.push(task);
+        }
+        let results = match futures::future::try_join_all(tasks).await {
+            Ok(res) => res,
+            Err(err) => panic!("Error processing transaction batches: {:?}", err),
+        };
+        // If any uploading fails, retry.
+        if any(results, |x| x.is_err()) {
+            anyhow::bail!("Uploading transactions failed.");
+        }
+
+        self.update_file_store_metadata(chain_id, start_version + batch_size as u64)
```

Comment:
no math. let's get the last element then add 1.

Reply:
Going to address this in the follow-up PR.